crates/partition-store/src/vqueue_table/waiting_reader.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2023 - 2025 Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use anyhow::Context;
+use rocksdb::DBRawIteratorWithThreadMode;
+
+use restate_storage_api::StorageError;
+use restate_storage_api::vqueue_table::{EntryCard, Stage, VQueueCursor};
+use restate_types::vqueue::VQueueId;
+
+use crate::PartitionDb;
+use crate::keys::{TableKey, TableKeyPrefix};
+use crate::vqueue_table::InboxKey;
+
+pub struct VQueueWaitingReader {
+    it: DBRawIteratorWithThreadMode<'static, rocksdb::DB>,
+}
+
+impl VQueueWaitingReader {
+    pub(crate) fn new(db: &PartitionDb, qid: &VQueueId) -> Self {
+        let mut readopts = rocksdb::ReadOptions::default();
+        readopts.set_async_io(true);
+        // this is not the place to be concerned about corruption, we favor speed
+        // over safety for this particular use-case.
+        readopts.set_verify_checksums(false);
+        readopts.set_tailing(true);
+        // Do not remove this!
+        readopts.set_total_order_seek(true);
+
+        // we know how big the prefix is
+        let mut key_buf = [0u8; InboxKey::by_stage_prefix_len()];
+        InboxKey::builder()
+            .partition_key(qid.partition_key)
+            .parent(qid.parent)
+            .instance(qid.instance)
+            .stage(Stage::Inbox)
+            .serialize_to(&mut key_buf.as_mut());
+
+        readopts.set_iterate_lower_bound(key_buf);
+        let success = super::convert_to_upper_bound(&mut key_buf);
+        debug_assert!(success);
+        readopts.set_iterate_upper_bound(key_buf);
+
+        let it = db
+            .rocksdb()
+            .inner()
+            .as_raw_db()
+            .raw_iterator_cf_opt(db.cf_handle(), readopts);
```

Comment:
For my understanding: The created iterator will pin a certain snapshot in time, right? So whenever there is a new inbox event that enqueues after the current head, we need to refresh/recreate the iterator, right?